theano/tensor/nnet/tests/test_nnet.py

```diff
@@ -1684,3 +1685,15 @@ def test_elu():
     for alpha in 1.5, 2, -1, -1.5, -2:
         y = elu(x, alpha).eval({x: X})
         utt.assert_allclose(y, numpy.where(X > 0, X, alpha * (numpy.exp(X) - 1)))
+
+
+def test_binary_crossentropy_reshape():
+    # Reported as https://github.com/Theano/Theano/issues/4086
+    a = tensor.tensor4('a')
+    c = binary_crossentropy(sigmoid(a.reshape((-1, 1))), 1).sum()
+    ga = theano.grad(c, a)
```

Comment:
I would add as a test the second case that had problems:

T.grad(binary_crossentropy(sigmoid(a).reshape((-1, 1)), 1).sum(), a).eval({a: numpy.array([[[[30.]]]])})